### DISCUSSION TITLE
HOGWARTS: Form will not submit. Uncaught TypeError: Cannot read property 'addEventListener' of null at buttonEvents

### DISCUSSION THREAD

**Discussion by katherinevfry** [nss-evening-cohort-14/evening-client-side#61] [Help Me!] [answered]
Your Name: 
Katy Fry

Have you searched the other help tickets? If not, do that first (Duplicates asks for help will be shamed ceaselessly)
Yep!

Describe your issue
I have created a function that handles all button events. My first button, which generates the input form, works fine. My second button, which submits the form info, keeps throwing a JS error in the console.

List any error messages that appear
main.js:15 Uncaught TypeError: Cannot read property 'addEventListener' of null
    at buttonEvents 

What have you googled? 
https://stackoverflow.com/questions/54171545/how-to-fix-cannot-read-property-addeventlistener-of-null-error
https://careerkarma.com/blog/cannot-read-property-addeventlistener-of-null/
https://forum.freecodecamp.org/t/targeting-dynamically-generated-id/261696/3

What have you tried? Tell us everything you have tried.
 I'm calling the correct ID. Currently, instead of a callback function, I'm using an anonymous function when I click the button to at least see if it's working before moving on to capturing the data from that click event. 

I don't call the button events function until the very end of my JS file, when I call it in an init function. However, it seems as if the DOM hasn't fully rendered when I click the submit button. I've tried moving the function further down in the file, but that doesn't seem to make a difference. I think the issue lies in the Id for the submit button being generated when the show form button is clicked, which is after the page has loaded. 

Copy/paste link TO THE FILE on your branch or in your repo that you are having issues with
https://github.com/katherinevfry/sorting-hat/blob/functions/main.js#L13

**Comment by drteresavasquez** [comment]
Ahh. So you're dynamically rendering the form. Nice! You cannot attach an event listener until the element is on the DOM, so you can either move your query selector for the the form button AFTER your function to get the form on the DOM INSIDE your showForm function OR you can target the parent and listen for a click on the button you created inside your button event function.

**Comment by drteresavasquez** [comment]
@katherinevfry let us know if you've figured this out. If so, can you put what your solution as a comment to your ticket? If not, we can keep digging.

**Comment by katherinevfry** [comment, marked as answer]
So I added my button event listener after the form and it works!

```
const showForm = () => {
  let form = `<form>
  <div>
    <label for="studentName">Student Name:</label>
    <input type="text" class="form-control" id="studentNameInput" placeholder="Nevil Longbottom" required>
  </div>
  <button type="submit" class="btn btn-primary" id="sortButton">Sort me!</button>
</form>`
printToDom ('#formDiv', form)
document.querySelector('#sortButton').addEventListener('click', function() {console.log('hi')});
  
}

```

My issue now is that the page is reloading automatically as soon as I submit any input. I know I need to use `e.preventDefault();` somewhere in my code (likely in my getFormInfo function, which I don't have working just yet.) I am seeing 'hi' briefly on the console before the page resets, so at least I know the button is working!

**Reply by drteresavasquez** [reply]
Awesome! If you end up needing help on the other problem, create a new ticket, but all looks good here!
